### DISCUSSION TITLE
Parse log is not correct

### DISCUSSION THREAD

**Discussion by uglyliu** [vectordotdev/vector#17252] [Q&A] [answered]
# Help me please
# Architecture
- Filebeat-->Kafka-->Vector-->ES
# Version
- vector 0.28.1 (x86_64-unknown-linux-gnu ff15924 2023-03-06)
# Original logs collected by kafka
`{"@timestamp":"2023-04-25T07:37:54.450Z","@metadata":{"beat":"filebeat","type":"_doc","version":"7.10.0" },"ecs":{"version":"1.6.0"},"log":{"offset":92,"file":{"path":"/var/log/nginx/java.log" }},"message":"2023-04-25 ha ha hello","input":{"type":"log"},"fields":{"kafka_topic":"java-test"},"host ":{"name":"VM-20-8-centos"},"agent":{"id":"e0941af7-9260-46ea-8486-7740c5a2ddba","name":"VM-20-8- centos","type":"filebeat","version":"7.10.0","hostname":"VM-20-8-centos","ephemeral_id":"d7fe6b6e-9fbe-4f10-a8a1-af46f0e1e64c" }}
`
As far as this kind of log is concerned, I can parse it successfully at https://playground.vrl.dev/, but it will report a parsing failure when it arrives at vector. The parsing rules are as follows:

`. = parse_regex!(.message, r'^(?P<time>\d+-\d+-\d+) (?P<haha>.*)$')`

![image](https://user-images.githubusercontent.com/22983421/235312318-e46fb970-86e1-4ea0-a847-14c50f161bbd.png)
![image](https://user-images.githubusercontent.com/22983421/235312554-dbd82332-0a7b-4bb6-a3e0-1a3747a9bf8e.png)

My all config is as follows:

[sources.dummy_logs]
type = "kafka" # required
bootstrap_servers = "127.0.0.1:9092" # required
group_id = "test-vector" # required
key_field = "vector_key" # optional, no default
auto_offset_reset = "latest"
commit_interval_ms = 2000
topics = ["java-test"] # required

[transforms.parse_logs]
type = "remap"
inputs = ["dummy_logs"]
drop_on_error = true
source        = '''
      . = parse_regex!(.message, r'^(?P<time>\d+-\d+-\d+) (?P<haha>.*)$')
'''

[sinks.elasticsearch]
 type = "elasticsearch" # required
 inputs = ["parse_logs"] # required
 api_version = "auto"
 compression = "none" # optional, default
 endpoints = ["http://127.0.0.1:9200"] # required
 mode = "bulk"
 bulk.action = "index"
 bulk.index = "vector-java-test-%Y.%m.%d" # optional, default
 #healthcheck.enabled = true # optional, default

**Comment by spencergilbert** [comment]
Can you please share the full error logs, rather than the shortened lines included in your image?

**Reply by uglyliu** [reply]
> Can you please share the full error logs, rather than the shortened lines included in your image?
`2023-05-01T14:00:11.535046Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vector::internal_events::remap: Mapping failed with event. error="function call error for \"parse_regex\" at (13:79): could not find any pattern matches" error_type="conversion_failed" stage="processing" internal_log_rate_limit=true
2023-05-01T14:00:11.535087Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vector_common::internal_event::component_events_dropped: Events dropped intentional=false count=1 reason="Mapping failed with event." internal_log_rate_limit=true
2023-05-01T14:00:24.260334Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vector::internal_events::remap: Mapping failed with event. error="function call error for \"parse_regex\" at (13:79): could not find any pattern matches" error_type="conversion_failed" stage="processing" internal_log_rate_limit=true
2023-05-01T14:00:24.260374Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vector_common::internal_event::component_events_dropped: Events dropped intentional=false count=1 reason="Mapping failed with event." internal_log_rate_limit=true
2023-05-01T14:00:24.263339Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vector::internal_events::remap: Internal log [Mapping failed with event.] is being rate limited.
2023-05-01T14:00:24.263368Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vector_common::internal_event::component_events_dropped: Internal log [Events dropped] is being rate limited.
`

**Reply by uglyliu** [reply]
> Can you please share the full error logs, rather than the shortened lines included in your image?

I want to realize the same function as logstash, cut off a log according to the correct rule, and then save it in es, so that it can be filtered according to the data segment

**Reply by spencergilbert** [reply]
It looks like you're getting unexpected content in your `message` then, I'd suggest adding logging to your parse function to better understand what's being received.

```toml
[transforms.parse_logs]
type = "remap"
inputs = ["dummy_logs"]
drop_on_error = true
source = '''
parsed, err = parse_regex(.message, r'^(?P\d+-\d+-\d+) (?P.*)$')
if err != null {
  log(.message, level: "error")
  abort  
}
. = parsed
'''
```

I'd also note your existing `. = parse_regex!()` would replace the entire event with the parsed contents when successful, dropping all of the existing fields on the event. You may want to use `merge()` if that's not desired.

**Reply by uglyliu** [reply]
> It looks like you're getting unexpected content in your `message` then, I'd suggest adding logging to your parse function to better understand what's being received.
> 
> ```toml
> [transforms.parse_logs]
> type = "remap"
> inputs = ["dummy_logs"]
> drop_on_error = true
> source = '''
> parsed, err = parse_regex(.message, r'^(?P\d+-\d+-\d+) (?P.*)$')
> if err != null {
>   log(err, level: "error")
>   abort  
> }
> . = parsed
> '''
> ```
> 
> I'd also note your existing `. = parse_regex!()` would replace the entire event with the parsed contents when successful, dropping all of the existing fields on the event. You may want to use `merge()` if that's not desired.

I try to run it,error is follows:

May 01 22:55:10 VM-20-8-centos vector[25197]: 2023-05-01T14:55:10.676019Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vrl_stdlib::log: function call error for "parse_regex" at (17:79): could not find any pattern matches internal_log_rate_secs=1 vrl_position=104
May 01 22:55:10 VM-20-8-centos vector[25197]: 2023-05-01T14:55:10.677760Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vrl_stdlib::log: Internal log [function call error for "parse_regex" at (17:79): could not find any pattern matches] is being rate limited.

**Reply by spencergilbert** [reply]
Apologies, copy paste mistake:

```toml
[transforms.parse_logs]
type = "remap"
inputs = ["dummy_logs"]
drop_on_error = true
source = '''
parsed, err = parse_regex(.message, r'^(?P\d+-\d+-\d+) (?P.*)$')
if err != null {
  log(.message, level: "error")
  abort  
}
. = parsed
'''
```

**Comment by uglyliu** [comment]
> parse_regex!(.message, r'^(?P<time>\d+-\d+-\d+) (?P<haha>.*)$')

Log is as follows:
May 01 23:09:10 VM-20-8-centos vector[30132]: 2023-05-01T15:09:10.683953Z ERROR transform{component_kind="transform" component_id=parse_logs component_type=remap component_name=parse_logs}: vrl_stdlib::log: {"@timestamp":"2023-05-01T15:09:09.679Z","@metadata":{"beat":"filebeat","type":"_doc","version":"7.10.0"},"log":{"offset":0,"file":{"path":"/etc/vector/java.log"}},"message":"2023-05-01 log log log","input":{"type":"log"},"fields":{"kafka_topic":"java-test"},"ecs":{"version":"1.6.0"},"host":{"name":"VM-20-8-centos"},"agent":{"version":"7.10.0","hostname":"VM-20-8-centos","ephemeral_id":"e92f14c8-1342-4c4a-8a6d-64fcfc26837c","id":"e0941af7-9260-46ea-8486-7740c5a2ddba","name":"VM-20-8-centos","type":"filebeat"}} internal_log_rate_secs=1 vrl_position=110

**Reply by spencergilbert** [reply]
Well, looking at your output and testing locally - you haven't provided the configuration you're actually using. The configuration you provided in the original question is invalid:

```shell
❯ vector validate test.toml
√ Loaded ["test.toml"]
2023-05-01T15:17:48.995753Z  WARN vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 61) error_type="request_failed" stage="processing" internal_log_rate_limit=true
2023-05-01T15:17:48.997449Z  WARN vector::sinks::elasticsearch::common: Failed to determine Elasticsearch version from `/_cluster/state/version`. Please fix the reported error or set an API version explicitly via `api_version`. assumed_version=8 error=Failed to get Elasticsearch API version: Failed to make HTTP(S) request: error trying to connect: tcp connect error: Connection refused (os error 61)
2023-05-01T15:17:49.325795Z ERROR rdkafka::client: librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): 127.0.0.1:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 1ms in state CONNECT)
2023-05-01T15:17:49.325862Z ERROR rdkafka::client: librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 1/1 brokers are down

Component errors
----------------
x Transform "parse_logs":
error[E101]: invalid regular expression
  ┌─ :1:28
  │
1 │ . = parse_regex!(.message, r'^(?P\d+-\d+-\d+) (?P.*)$')
  │                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^ regex parse error: unrecognized flag: ^(?P\d+-\d+-\d+) (?P.*)$
  │
  = see documentation about regular expressions at https://vrl.dev/expressions/#regular-expression
  = learn more about error code 101 at https://errors.vrl.dev/101
  = see language documentation at https://vrl.dev
  = try your code in the VRL REPL, learn more at https://vrl.dev/examples
  ```
  
Please confirm what you're actually running and share the correct configuration for more assistance, thanks!

**Reply by uglyliu** [reply]
> Well, looking at your output and testing locally - you haven't provided the configuration you're actually using. The configuration you provided in the original question is invalid:
> 
> ```shell
> ❯ vector validate test.toml
> √ Loaded ["test.toml"]
> 2023-05-01T15:17:48.995753Z  WARN vector::internal_events::http_client: HTTP error. error=error trying to connect: tcp connect error: Connection refused (os error 61) error_type="request_failed" stage="processing" internal_log_rate_limit=true
> 2023-05-01T15:17:48.997449Z  WARN vector::sinks::elasticsearch::common: Failed to determine Elasticsearch version from `/_cluster/state/version`. Please fix the reported error or set an API version explicitly via `api_version`. assumed_version=8 error=Failed to get Elasticsearch API version: Failed to make HTTP(S) request: error trying to connect: tcp connect error: Connection refused (os error 61)
> 2023-05-01T15:17:49.325795Z ERROR rdkafka::client: librdkafka: Global error: BrokerTransportFailure (Local: Broker transport failure): 127.0.0.1:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused (after 1ms in state CONNECT)
> 2023-05-01T15:17:49.325862Z ERROR rdkafka::client: librdkafka: Global error: AllBrokersDown (Local: All broker connections are down): 1/1 brokers are down
> 
> Component errors
> ----------------
> x Transform "parse_logs":
> error[E101]: invalid regular expression
>   ┌─ :1:28
>   │
> 1 │ . = parse_regex!(.message, r'^(?P\d+-\d+-\d+) (?P.*)$')
>   │                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^ regex parse error: unrecognized flag: ^(?P\d+-\d+-\d+) (?P.*)$
>   │
>   = see documentation about regular expressions at https://vrl.dev/expressions/#regular-expression
>   = learn more about error code 101 at https://errors.vrl.dev/101
>   = see language documentation at https://vrl.dev
>   = try your code in the VRL REPL, learn more at https://vrl.dev/examples
> ```
> 
> Please confirm what you're actually running and share the correct configuration for more assistance, thanks!

It is a problem with the github page, that character is not displayed，I provide my configuration

 
![image](https://user-images.githubusercontent.com/22983421/235476825-bdad787a-1b0c-42bd-bc77-c959e1799a9b.png)

**Reply by spencergilbert** [reply]
You can provide your configurations in [code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks), by wrapping the contents with triple `s.

```coffeescript
parsed, err = parse_regex(.message, r'^(?P<time>\d+-\d+-\d+) (?P<haha>.*)$')
if err != null {
  log(., level: "error")
  abort  
}
. = parsed
```

It seems likely your event is not structured as you expect.

**Comment by uglyliu** [comment]
![Uploading image.png…]()

**Comment by uglyliu** [comment]
[transforms.parse_logs]
type = "remap"
inputs = ["dummy_logs"]
drop_on_error = true
source        = '''
      parsed, err = parse_json(.message)
      if err != null {
      log(.message, level: "error")
      abort  
      }
     . = parsed
'''
this config can play, But it's not the effect ，I want to “log log log” is a field, “2023-05-01” is a other field，rather than together

![image](https://user-images.githubusercontent.com/22983421/235478880-0992d0d2-4c66-43a9-80fd-76c866518a2c.png)

**Reply by spencergilbert** [reply]
Because you used `parse_json()` which failed, as it isn't JSON and thus your program aborted.

**Reply by uglyliu** [reply]
> Because you used `parse_json()` which failed, as it isn't JSON and thus your program aborted.

Excuse me, how can I achieve the effect I want?

**Comment by uglyliu** [comment]
> {"@timestamp":"2023-04-25T07:37:54.450Z","@metadata":{"beat":"filebeat","type":"_doc","version":"7.10.0" },"ecs":{"version":"1.6.0"},"log":{"offset":92,"file":{"path":"/var/log/nginx/java.log" }},"message":"2023-04-25 ha ha hello","input":{"type":"log"},"fields":{"kafka_topic":"java-test"},"host ":{"name":"VM-20-8-centos"},"agent":{"id":"e0941af7-9260-46ea-8486-7740c5a2ddba","name":"VM-20-8- centos","type":"filebeat","version":"7.10.0","hostname":"VM-20-8-centos","ephemeral_id":"d7fe6b6e-9fbe-4f10-a8a1-af46f0e1e64c" }}

I solved it with parse_grok，my config is as follows
```
[transforms.parse_logs]
type = "remap"
inputs = ["dummy_logs"]
drop_on_error = true
source        = """
            .message = "2023-01-04 14:03:30.745 blah blah blah blah"
            .flat = parse_grok!(.message, "^%{TIMESTAMP_ISO8601:timestamp} %{GREEDYDATA:logline}")
            .log = .flat.logline
            .currentdate = .flat.timestamp
           del(.flat)
"""

```
There is one last problem that has not been solved, how to convert 2023-01-04 14:03:30.745 to 2023.01.04
![image](https://user-images.githubusercontent.com/22983421/235551543-8dc6d6c5-4aa9-4ef1-acda-f3f6d71712aa.png)

**Reply by spencergilbert** [reply]
You can leverage [format_timestamp()](https://vector.dev/docs/reference/vrl/functions/#format_timestamp).

**Reply by uglyliu** [reply, marked as answer]
> You can leverage [format_timestamp()](https://vector.dev/docs/reference/vrl/functions/#format_timestamp).

thank u very much, it is solved ,my all config is as follows

```
data_dir = "/var/lib/vector"
#[sources.dummy_logs]
#type = "file"
#include = [ "/etc/vector/java.log" ]
[sources.dummy_logs]
type = "kafka" # required
bootstrap_servers = "127.0.0.1:9092" # required
group_id = "test-vector" # required
key_field = "vector_key" # optional, no default
auto_offset_reset = "latest"
commit_interval_ms = 2000
topics = ["java-test"] # required
#topics = ["nginx-test"] # required

# Parse Syslog logs
# See the Vector Remap Language reference for more info: https://vrl.dev
[transforms.parse_logs]
type = "remap"
inputs = ["dummy_logs"]
drop_on_error = true
source        = """
            .message = "2023-01-04 14:03:30.745 blah blah blah blah"
            .flat = parse_grok!(.message, "^%{TIMESTAMP_ISO8601:timestamp} %{GREEDYDATA:logline}")
            .log = .flat.logline
            .currentdate = format_timestamp!(now(), format: "%Y.%m.%d")
            del(.flat)    

"""




# Print parsed logs to stdout
#[sinks.print]
#type = "console"
#inputs = ["parse_logs"]
#encoding.codec = "json"
[sinks.elasticsearch]
 type = "elasticsearch" # required
 inputs = ["parse_logs"] # required
 api_version = "auto"
 compression = "none" # optional, default
 endpoints = ["http://127.0.0.1:9200"] # required
 mode = "bulk"
 bulk.action = "index"
 bulk.index = "vector-yauto-{{ .currentdate }}" # optional, default

 #healthcheck.enabled = true # optional, default

```

![image](https://user-images.githubusercontent.com/22983421/235709143-3d765fd9-8745-4a11-888a-914fbbab06f4.png)
